packages/gatsby-theme-store/src/sdk/toast/Provider.tsx

```diff
@@ -0,0 +1,70 @@
+import React, {
+  createContext,
+  useCallback,
+  useEffect,
+  useRef,
+  useState,
+} from 'react'
+import type { FC } from 'react'
+
+export interface ShowToastOptions {
+  message: string
+  duration?: number
+}
+
+export interface State {
+  isVisible: boolean
+  message: string
+}
+
+export interface IContext extends State {
+  showToast: (options: ShowToastOptions) => void
+  hideToast: () => void
+}
+
+export const Context = createContext<IContext | undefined>(undefined)
+
+export const Provider: FC = ({ children }) => {
+  const [toastState, setToastState] = useState<State>({
+    isVisible: false,
+    message: '',
+  })
+
+  const timeoutRef = useRef(0)
+
+  useEffect(() => {
+    return () => clearTimeout(timeoutRef.current)
+  }, [])
+
+  const hideToast = useCallback(() => {
+    setToastState((prev) => ({ ...prev, isVisible: false }))
+  }, [])
+
+  const showToast = useCallback<IContext['showToast']>(
+    ({ duration = 3000, message }: ShowToastOptions) => {
+      setToastState((prevState) => ({
+        ...prevState,
+        message,
+        isVisible: true,
+      }))
+
+      timeoutRef.current = (setTimeout(
+        () => setToastState((prev) => ({ ...prev, isVisible: false })),
+        duration
+      ) as unknown) as number
+    },
+    []
+  )
+
+  return (
+    <Context.Provider
+      value={{
+        showToast,
+        hideToast,
+        ...toastState,
+      }}
```

Comment:
Add useMemo to toast context value

Reply:
I think isn't required, because you will use each property from the context, that is a function with usecallback or a primary value from state.

Reply:
Everyone that uses useToast() will re-render when this object changes. If that's expected ok.